### PR TITLE
Feat: OpenShell3D.project coincident faces

### LINKED ISSUE #447: Footprint of one closed shell on another closed shell
* **I'm submitting a ...**
  - [ ] bug report
  - [X] feature request

* **What is the current behavior?**

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem** Avoid reference to other packages

* **What is the expected behavior?**
Returns if exists (at least 2 coincident faces between the 2 closed shells) the closed shell 1 with the footprint of the closed shell 2.

* **What is the motivation / use case for changing the behavior?**

* **Possible fixes**

* **Please tell us about your environment:**
  
  - branch: 
  - commit:  
  - python version:

### PR DESCRIPTION
**Pull request by sbendjebla**
* **Please check if the PR fulfills these requirements**
- [ ] The commit messages follows our guidelines (explicit, in english)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] CHANGELOG.md has been updated

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [ ] Bug fix:
- [ ] New features:
- [ ] Documentation
- [ ] Optimization

* **Other information**:
Divides the 1st `OpenShell3D` faces based on coincident faces of the 2nd `OpenShell3D`, so as contours on both shells are superposing

Closes #447 
Wait for https://github.com/Dessia-tech/volmdlr/pull/544 to be merged

### CODE REVIEW

**Review by masfaraud — APPROVED** [review]

File: volmdlr/faces.py [hunk 1]
```diff
@@ -6982,6 +6982,41 @@ def plot(self, ax=None, color: str = 'k', alpha: float = 1):
...
+    def project_coincident_faces_of(self, shell):
+        """
+        Divides self's faces based on coincident shell's faces
+        """
+
+        used_faces, list_faces = {}, []
+        initial_faces = self.faces[:]
+
+        for _, face1 in enumerate(initial_faces):
+            for _, face2 in enumerate(shell.faces):
```
Comment (WirajanDASILVA) [inline volmdlr/faces.py:6994]: Here since you're not using the indexes from enumerate, could you remove them, please?

Comment (WirajanDASILVA) [inline volmdlr/faces.py:6994]:
```suggestion
        for face1 in initial_faces:
            for face2 in shell.faces:
```

**Review by WirajanDASILVA — CHANGES_REQUESTED** [review]

File: volmdlr/faces.py (hunk 1 above)
Comment (sbendjebla) [reply volmdlr/faces.py]: I've added your commit ! Thanks !

**Review by WirajanDASILVA — APPROVED** [review]

File: volmdlr/faces.py [hunk 2]
```diff
@@ -7557,6 +7561,138 @@ def plot(self, ax=None, color: str = 'k', alpha: float = 1):
...
+        """
+        Divides self's faces based on coincident shell's faces
+        """
+
+        used_faces, list_faces = {}, []
+        initial_faces = self.faces[:]
+
+        for i, face1 in enumerate(initial_faces):
+            for j, face2 in enumerate(shell.faces):
+                contour1 = face1.surface2d.outer_contour
```
Comment (masfaraud) [inline volmdlr/faces.py:7574]: Could we make this part of the code a method of face taking another face into parameter?

**Review by masfaraud — CHANGES_REQUESTED** [review]

File: volmdlr/faces.py (hunk 2 above)
Comment (sbendjebla) [reply volmdlr/faces.py]: @masfaraud 
I tried to define a new method for PlaneFace3D which takes a list of faces (it is more significant than just one face) -> here in this [**commit**](https://github.com/Dessia-tech/volmdlr/pull/463/commits/5f4350fe93bb1b8f23a9d7257f5082d128a7d8c2)

File: volmdlr/faces.py [hunk 3]
```diff
@@ -5237,6 +5241,138 @@ def cut_by_coincident_face(self, face):
...
+                    #                 list_faces.append(divided_faces_d_face[0])
+                    #                 d_face = divided_faces_d_face[1]
+
+                    #             if divided_faces_d_face:
+                    #                 list_faces.append(divided_faces_d_face[1])
+
+                    #         else:
+                    #             list_faces.append(d_face)
+                    #     else:
+                    #         used.append(d_face)
```
Comment (masfaraud) [inline volmdlr/faces.py:5360]: Cleanup?

File: volmdlr/faces.py [hunk 4]
```diff
@@ -5237,6 +5241,138 @@ def cut_by_coincident_face(self, face):
...
+
+                    used, list_faces = self.update_faces_with_divided_faces(
+                        divided_faces, face2_2, used, list_faces)
+                used_faces[self] = used
+
+        try:
+            if isinstance(used_faces[self], list):
```
Comment (masfaraud) [inline volmdlr/faces.py:5367]: Why is the check needed? used_faces can either be a list or a dict?

File: volmdlr/faces.py [hunk 5]
```diff
@@ -7571,6 +7707,151 @@ def plot(self, ax=None, color: str = 'k', alpha: float = 1):
...
+    #     c_inners_2 = [face1.surface3d.contour3d_to_2d(inner) for inner in face2.inner_contours3d]
+    #     inside = set()
+    #     for c1 in c_inners_1:
+    #         for c2 in c_inners_2:
+    #             if c1.is_superposing(c2):
+    #                 inside.add(False)
+    #             else:
+    #                 inside.add(c2.is_inside(c1))
+    #     return inside
+
```
Comment (masfaraud) [inline volmdlr/faces.py:7854]: Clean

**Review by masfaraud — CHANGES_REQUESTED** [review]

**Review by masfaraud — APPROVED** [review]

**Pull request merged**